examples/ipfs-private/src/main.rs

```diff
@@ -118,76 +95,72 @@ async fn main() -> Result<(), Box<dyn Error>> {
-
     // Create a Gosspipsub topic
     let gossipsub_topic = gossipsub::IdentTopic::new("chat");
 
     // We create a custom network behaviour that combines gossipsub, ping and identify.
     #[derive(NetworkBehaviour)]
-    #[behaviour(to_swarm = "MyBehaviourEvent")]
     struct MyBehaviour {
         gossipsub: gossipsub::Behaviour,
         identify: identify::Behaviour,
         ping: ping::Behaviour,
     }
 
-    enum MyBehaviourEvent {
-        Gossipsub(gossipsub::Event),
-        Identify(identify::Event),
-        Ping(ping::Event),
-    }
-
-    impl From<gossipsub::Event> for MyBehaviourEvent {
-        fn from(event: gossipsub::Event) -> Self {
-            MyBehaviourEvent::Gossipsub(event)
-        }
-    }
-
-    impl From<identify::Event> for MyBehaviourEvent {
-        fn from(event: identify::Event) -> Self {
-            MyBehaviourEvent::Identify(event)
-        }
-    }
-
-    impl From<ping::Event> for MyBehaviourEvent {
-        fn from(event: ping::Event) -> Self {
-            MyBehaviourEvent::Ping(event)
-        }
-    }
-
-    // Create a Swarm to manage peers and events
-    let mut swarm = {
-        let gossipsub_config = gossipsub::ConfigBuilder::default()
-            .max_transmit_size(262144)
-            .build()
-            .expect("valid config");
-        let mut behaviour = MyBehaviour {
-            gossipsub: gossipsub::Behaviour::new(
-                gossipsub::MessageAuthenticity::Signed(local_key.clone()),
-                gossipsub_config,
-            )
-            .expect("Valid configuration"),
-            identify: identify::Behaviour::new(identify::Config::new(
-                "/ipfs/0.1.0".into(),
-                local_key.public(),
-            )),
-            ping: ping::Behaviour::new(ping::Config::new()),
-        };
-
-        println!("Subscribing to {gossipsub_topic:?}");
-        behaviour.gossipsub.subscribe(&gossipsub_topic).unwrap();
-        SwarmBuilder::with_async_std_executor(transport, behaviour, local_peer_id).build()
-    };
+    let mut swarm = libp2p::SwarmBuilder::with_new_identity()
+        .with_async_std()
+        .with_other_transport(|key| {
+            let noise_config = noise::Config::new(key).unwrap();
+            let yamux_config = yamux::Config::default();
+
+            let base_transport =
+                tcp::async_io::Transport::new(tcp::Config::default().nodelay(true));
+            let maybe_encrypted = match psk {
+                Some(psk) => Either::Left(
+                    base_transport
+                        .and_then(move |socket, _| PnetConfig::new(psk).handshake(socket)),
+                ),
+                None => Either::Right(base_transport),
+            };
+            maybe_encrypted
+                .upgrade(Version::V1Lazy)
+                .authenticate(noise_config)
+                .multiplex(yamux_config)
+                .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+        })?
```

Comment:
While _basetransport_ is easy to grasp for us maintainers, familiar with the upgrade flow of TCP, I don't think the same is true for regular users. I.e. I don't think they would know that TCP is a basetransport whereas QUIC is a full transport. Thus I don't think they would know that they have to call `with_base_transport` with a TCP transport.

Next not all basetransports are equal. While TCP can be placed anywhere within the hierarchy, websocket has to come last, wrapping a DNS and a TCP transport. Thus I don't think the `with_base_transport` abstraction is solid.

Does that make sense @thomaseizinger?

Reply:
Sounds good. This comment predates our synchronous conversation about this where I think we agreed that we are going to try something like:

```rust
fn with_tcp(self, security_protocols: [TSec; NUM_SEC], muxer_protocols: [TMux; NUM_MUX])
``

When typing out this API sketch, I noticed that arrays are maybe not such a good idea here because they aren't heterogeneous. Perhaps we should do something like:

```rust
fn with_tcp(self, security_protocols: impl IntoSec, muxer_protocols: impl IntoMuxer);

trait IntoSec {
	fn into_sec(self) -> Upgrade;
}

impl<T1> IntoSec for T1 { }
impl<T1, T2> IntoSec for (T1, T2) { }
// etc
```

That would allow the user to call it with tuples of different types. Otherwise we might have to introduce a boxed upgrade and calling the API isn't quite as convenient. For now, I'd suggest we implement it for up to 2 elements because that is how many muxers & security protocols we have. If more appear, users can always open an issue.